content/guides/typescript-for-solid.mdx

```diff
@@ -0,0 +1,786 @@
+
+### `ref` attribute
+
+#### Basics
+
+In an environment without TypeScript, using the `ref` attribute in Solid ensures that the corresponding DOM element is assigned to the variable after it is rendered.:
+
+```js
+let divRef;
+console.log(divRef); // Outputs: undefined
+onMount(() => {
+  console.log(divRef); // Outputs: <div> element
+});
+return <div ref={divRef} />;
+```
+
+In a TypeScript environment, particularly with strict `null` checks enabled, typing these variables can be a challenge.
+
+A safe approach in TypeScript is to acknowledge that `divRef` may initially be `undefined` and to implement checks when accessing it:
+
+```ts
+let divRef: HTMLDivElement | undefined;
+// This would be flagged as an error during compilation
+divRef.focus(); 
+onMount(() => {
+  if (!divRef) return;
+  divRef.focus(); 
+});
+return <div ref={divRef}>...</div>;
+```
+
+Within the scope of the `onMount` function, which runs after rendering, a non-`null` assertion (indicated by the exclamation mark `!`) can be used:
+
+```typescript
+onMount(() => {
+  divRef!.focus();
+});
+```
+
+Another approach is to bypass `null` during the assignment phase and then apply a definite assignment assertion within the `ref` attribute:
+
+```typescript
+let divRef: HTMLDivElement;
+// Compilation error as expected
+divRef.focus(); 
+onMount(() => {
+  divRef.focus(); 
+});
+return <div ref={divRef!}>...</div>;
+```
+
+In this case, using `divRef!` within the `ref` attribute signals to TypeScript that `divRef` will receive an assignment after this stage, which is more in line with how Solid works.
+
+<Aside>
+  **Caution**: 
+  While TypeScript does catch incorrect usage of refs that occur before their JSX block definition, it currently does not identify undefined variables within certain nested functions in Solid. 
+  Therefore, additional care is needed when using `ref`s in functions such as [`createMemo`](/reference/createMemo), [`createRenderEffect`](/reference/createRenderEffect), and [`createComputed`](/reference/createComputed).
+</Aside>
+
+Finally, a riskier approach involves using the definite assignment assertion right at the point of variable initialization. 
+While this method bypasses TypeScript's assignment checks for that particular variable, it offers a quick but less secure workaround that could lead to runtime errors.
+
+```typescript
+let divRef!: HTMLDivElement;
+
+// Permitted by TypeScript but will throw an error at runtime:
+// divRef.focus();
+
+onMount(() => {
+  divRef.focus(); 
+});
+```
+
+## Control flow based narrowing
+
+Control flow-based narrowing involves refining the type of a value by using control flow statements.
+
+Consider the following:
+
+```tsx
+const user: User | undefined = maybeUser();
+return <div>{user && user.name}</div>;
+```
+
+In Solid, however, Accessors cannot be narrowed in a similar way:
```

Comment:
> Accessors

Are accessors a proper noun or was the capitalization an error? Because you don't capitalize later.